raft/server.cc

```diff
@@ -449,9 +450,13 @@ future<> server_impl::wait_for_entry(entry_id eid, wait_type type, seastar::abor
 }
 
 future<entry_id> server_impl::add_entry_on_leader(command cmd, seastar::abort_source* as) {
+    const auto memory_usage = memory_usage_of(cmd);
+    if (const auto limit = _config.max_log_size_in_bytes - _config.max_snapshot_trailing_bytes; memory_usage > limit) {
+        throw std::runtime_error(fmt::format("command memory usage {} is greater than limit {}", memory_usage, limit));
+    }
     // Wait for a new slot to become available
     try {
-        co_await _fsm->wait_max_log_size(as);
+        co_await _fsm->consume_memory(as, memory_usage);
```

Comment:
I think we have a pre-existing problem here but it made worse by the patch. What if `fsm->add_entry` throws? Semaphore units are not freed.

Reply:
I'll write write a test for it and add a try/catch

Reply:
done